src/Stratis.Bitcoin.Features.BlockStore/BlockStoreSignaled.cs

```diff
@@ -125,84 +173,75 @@ protected override void OnNextCore(Block block)
 
-                if (!broadcastItems.Any())
+                // Check if we've reorged away from the current block.
+                if (this.chainState.ConsensusTip.FindAncestorOrSelf(block) == null)
                 {
-                    this.logger.LogTrace("(-)[NO_BROADCAST_ITEMS]");
-                    return;
-                }
+                    this.logger.LogTrace("Block header '{0}' not found in the consensus chain.", block);
 
-                IReadOnlyNetworkPeerCollection peers = this.connection.ConnectedPeers;
-                if (!peers.Any())
-                {
-                    this.logger.LogTrace("(-)[NO_PEERS]");
-                    return;
+                    // Remove hash that we've reorged away from.
+                    this.localBatch.Remove(block);
+                    continue;
                 }
+               
+                this.logger.LogTrace("Block '{0}' moved from the announce queue to broadcast list.", block);
+                this.localBatch.Remove(block);
+                broadcastItems.Add(block);
+            }
+
+            if (!broadcastItems.Any())
+            {
+                this.logger.LogTrace("(-)[NO_BROADCAST_ITEMS]");
+                return;
+            }
 
-                // Announce the blocks to each of the peers.
-                IEnumerable<BlockStoreBehavior> behaviours = peers.Select(s => s.Behavior<BlockStoreBehavior>());
+            IReadOnlyNetworkPeerCollection peers = this.connection.ConnectedPeers;
+            if (!peers.Any())
+            {
+                this.logger.LogTrace("(-)[NO_PEERS]");
+                return;
+            }
 
-                this.logger.LogTrace("{0} blocks will be sent to {1} peers.", broadcastItems.Count, behaviours.Count());
-                foreach (BlockStoreBehavior behaviour in behaviours)
-                    await behaviour.AnnounceBlocksAsync(broadcastItems).ConfigureAwait(false);
-            },
-            this.nodeLifetime.ApplicationStopping,
-            repeatEvery: TimeSpans.Second,
-            startAfter: TimeSpans.FiveSeconds);
+            // Announce the blocks to each of the peers.
+            IEnumerable<BlockStoreBehavior> behaviours = peers.Select(s => s.Behavior<BlockStoreBehavior>());
 
-            this.logger.LogTrace("(-)");
+            this.logger.LogTrace("{0} blocks will be sent to {1} peers.", broadcastItems.Count, behaviours.Count());
+            foreach (BlockStoreBehavior behaviour in behaviours)
+                await behaviour.AnnounceBlocksAsync(broadcastItems).ConfigureAwait(false);
         }
 
-        /// <summary>
-        /// The async loop needs to complete its work before we can shut down this feature.
-        /// </summary>
-        internal void ShutDown()
+        /// <inheritdoc />
+        protected override void Dispose(bool disposing)
         {
-            this.asyncLoop.Dispose();
+            this.blocksToAnnounce.Dispose();
```

Comment:
We need to find a way to wait for the timer invoked method.

Reply:
we can do the disposing inside the async lock

Reply:
Hmmm interesting idea, but then you might not be the next in line to get the lock.
Ah and the AsyncLock will throw anyway as it uses cancellation if we wait for the async lock with a None cancellation that might work as any tasks that are ahead of the dispose method will just throw.
Not sure that will work.

Reply:
You're right, just checked that. 
So I guess I'll just add a local task that we can await on dispose.

Reply:
Fixed